### PR TITLE
[docs] training performance

### PR DESCRIPTION
**Pull request by stevhliu**
- created a new performance section divided into memory and speed optimizations
- model memory training anatomy [guide](https://huggingface.co/docs/transformers/main/en/model_memory_anatomy) is now the more descriptive and simplified GPU memory usage doc
- add brief loss scaling section in gradient accumulation for custom loss function users

### CODE REVIEW

File: docs/source/en/grad_accumulation.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+args = TrainingArguments(
+    ...,
+    per_device_train_batch_size=8,
+    gradient_accumulation_steps=4,
+)
+```
+
+## Loss scaling
+
+[`Trainer`] automatically divides loss by `gradient_accumulation_steps` before the backward pass. But if you're using a [custom loss function](./trainer_recipes#custom-loss-function) that accepts `num_items_in_batch` (total non-padding token count across all mini-batches), you need to handle the loss normalization yourself.
```
Comment (SunMarc) [inline docs/source/en/grad_accumulation.md]: > [`Trainer`] automatically divides loss by `gradient_accumulation_steps` before the backward pass

This was the case before but this isn't the case anymore. Basically, if the loss defined takes `num_items_in_batch`, it will correctly calculate the loss. Otherwise, it will just divide by `gradient_accumulation_steps. I think for the doc here, we can just show the user how to correctly define their loss e.g must use `num_items_in_batch` arg. 

If we want also, we can refer to unsloth blog to show how the grad acc is computed: https://unsloth.ai/blog/gradient

File: docs/source/en/grad_accumulation.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+def my_loss_fn(outputs, labels, num_items_in_batch=None):
+    logits = outputs["logits"]
+    if num_items_in_batch is not None:
+        loss = F.cross_entropy(logits, labels, reduction="sum")
+        loss = loss / num_items_in_batch                           # scale loss
+    else:
+        loss = F.cross_entropy(logits, labels, reduction="mean")
+    return loss
+```
```
Comment (SunMarc) [inline docs/source/en/grad_accumulation.md]: For now, if they don't compute their loss with `num_items_in_batch`, the results will wrong, so let's make it non optional.

```suggestion
def my_loss_fn(outputs, labels, num_items_in_batch):
    logits = outputs["logits"]
    loss = F.cross_entropy(logits, labels, reduction="sum")
    loss = loss / num_items_in_batch # scale loss
    return loss
```

File: docs/source/en/grad_checkpointing.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+> Use with [gradient accumulation](./grad_accumulation) to further reduce memory usage.
+
+```py
+from transformers import TrainingArguments
+
+args = TrainingArguments(
+    ...,
+    gradient_checkpointing=True,
+)
+```
```
Comment (SunMarc) [inline docs/source/en/grad_checkpointing.md:44]: Just to let you know, in fsdp, this is called `activation_checkpointing`. Both options are possible for fsdp but in trainer we will warn the user to use `activation_checkpointing` instead !

File: docs/source/en/kernels.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    kernel_mapping={
+        "RMSNorm": "kernels-community/rmsnorm",
+    }
+)
+model = AutoModelForCausalLM.from_pretrained(
+    "Qwen/Qwen3-0.6B",
+    use_kernels=True,
+    kernel_config=kernel_config,
+)
+```
```
Comment (SunMarc) [inline docs/source/en/kernels.md:38]: maybe we can show also the attention api here or in a new section as this also depends partially on hub kernels ? or redirect to the doc we already have ?

File: docs/source/en/kernels.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        "cross_entropy": True,
+        "rms_norm": False,
+        "swiglu": True,
+    }
+)
+```
+
+## Next steps
+
+- See the [Attention backends](./attention_interface) guide for details on kernels like FlashAttention that reduce memory usage.
```
Comment (SunMarc) [inline docs/source/en/kernels.md:81]: do we already have this doc ?

File: docs/source/en/grad_checkpointing.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+## Next steps
+
+- Read the [GPU memory usage](./model_memory_anatomy) doc to understand what is driving memory usage on the GPU during training.
+- See the [Mixed precision training](./mixed_precision_training) guide to learn how to use lower precision data types to further reduce memory and speed up training.
+- See the [Kernels](./kernels) guide to learn how to speed up training with custom fused kernels.
```
Comment (SunMarc) [inline docs/source/en/grad_checkpointing.md:50]: Instead of these, maybe it will be simpler to maintain to have an overview doc with all these techniques summarized ?

File: docs/source/en/mixed_precision_training.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+Set [`~TrainingArguments.bf16`] or [`~TrainingArguments.fp16`]  to `True` to enable mixed precision training. These are both 16-bit data types but bf16 has the same exponent range as fp32, so it almost never runs into overflow issues. Use bf16 on more modern hardware like A100, H100, or Ampere GPUs and fallback to fp16 on hardware like V100 or T4.
+
+```py
+from transformers import TrainingArguments
+
+args = TrainingArguments(..., bf16=True)
+args = TrainingArguments(..., fp16=True)
+```
```
Comment (SunMarc) [inline docs/source/en/mixed_precision_training.md]: Maybe we can add a warning saying that the user will need to load the model in float32 otherwise it becomes the `autocast` becomes a no-op ? 
Maybe also specify that the user can just directly load and train the model in bf16/fp16 if it is stable enough. dtype=torch.bfloat16 or torch.float16

File: docs/source/en/model_memory_anatomy.md [hunk 8]
```diff
@@ -14,255 +14,69 @@ See the License for the specific language governing permissions and
...
+│  (Adam)              ├── fp32 momentum  (4B)                         │
+│                      └── fp32 variance  (4B)                         │
+│                      ↳ quantized Adam (bitsandbytes) → 2 bytes/param │
+│                                                                      │
+│  GRADIENTS           ████████████████  4 bytes/param  (fp32)         │
+│                      └── computed in backward pass, per parameter    │
+│                                                                      │
+│  ACTIVATIONS         ████ varies — batch × seq_len × depth × hidden  │
+│  (forward cache)     └── cached for backward; can OOM even if model  │
+│                          fits (long seqs / large batches)            │
```
Comment (SunMarc) [inline docs/source/en/model_memory_anatomy.md:38]: these are in fp16/bf16 if using mixed precision

File: docs/source/en/torch_compile.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+[torch.compile](https://docs.pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) compiles PyTorch code to fused kernels to make it run faster. For training, it traces both the forward and backward pass together and compiles them into optimized kernels, reducing the overhead of individual op launches and fusing operations to cut memory bandwidth usage.
+
+The table below lists the backends that support training.
+
+| backend | description |
+|---|---|
+| `inductor` | default; traces forward/backward graphs ahead of time and compiles to Triton kernels that run on CUDA |
+| `aot_cudagraphs` | captures the GPU operation sequence and replays it with minimal CPU dispatch overhead; best for fixed-shape workloads |
+| `aot_nvfuser` / `nvfuser` | NVIDIA's fusion compiler via AOTAutograd or TorchScript |
+
```
Comment (SunMarc) [inline docs/source/en/torch_compile.md]: maybe show this instead

```
 Training & Inference backends:
  ┌────────────┬───────────────────────────────────┐
  │  Backend   │            Description            │
  ├────────────┼───────────────────────────────────┤
  │ inductor   │ TorchInductor backend (default)   │
  ├────────────┼───────────────────────────────────┤
  │ cudagraphs │ CUDA graphs with AOT Autograd     │
  ├────────────┼───────────────────────────────────┤
  │ ipex       │ Intel Extension for PyTorch (CPU) │
  └────────────┴───────────────────────────────────┘

 Mode: default
  Description: Balanced compile time vs runtime
  ────────────────────────────────────────
  Mode: reduce-overhead
  Description: Reduces Python/CPU overhead using CUDA
    graphs, at the cost of some extra memory
  ────────────────────────────────────────
  Mode: max-autotune
  Description: Benchmarks multiple kernel
  implementations
    at compile time, picks the fastest. Longer
    compilation
  ────────────────────────────────────────
  Mode: max-autotune-no-cudagraphs
  Description: Same as max-autotune but without CUDA
    graphs
```

**Review by SunMarc — COMMENTED** [review]
Really nice ! Thanks a lot, left a couple of comments but feel free to disregard some of them !

File: docs/source/en/kernels.md (hunk 4 above)
Comment (stevhliu) [reply docs/source/en/kernels.md:38]: yeah i think redirect to existing doc is better so we avoid duplicating content!

File: docs/source/en/kernels.md (hunk 5 above)
Comment (stevhliu) [reply docs/source/en/kernels.md:81]: yeah! see here: https://huggingface.co/docs/transformers/main/en/attention_interface

File: docs/source/en/grad_checkpointing.md (hunk 6 above)
Comment (stevhliu) [reply docs/source/en/grad_checkpointing.md:50]: i think its good to have this to connect users to other related topics instead of a dead end. but i agree we should do an overview doc similar to this [inference one](https://huggingface.co/docs/transformers/main/en/optimization_overview)

**Review by SunMarc — APPROVED** [review]
Much better thanks !
